File: apps/expo/modules/readium/ios/PDFView.swift
import ExpoModulesCore
import ReadiumNavigator
import ReadiumShared
import ReadiumAdapterGCDWebServer
import ReadiumInternal
import UIKit

public struct PDFProps {
    var bookId: String?
    var locator: Locator?
    var initialLocator: Locator?
    var url: String?
    var background: Color?
    var pageSpacing: Double?
    var scrollAxis: Axis?
    var scroll: Bool?
    var readingProgression: ReadiumNavigator.ReadingProgression?
    var spread: Spread?
}

public struct FinalizedPDFProps {
    var bookId: String
    var locator: Locator?
    var url: String
    var background: Color
    var pageSpacing: Double
    var scrollAxis: Axis
    var scroll: Bool
    var readingProgression: ReadiumNavigator.ReadingProgression
    var spread: Spread
}

public class PDFView: ExpoView {
    let onLocatorChange = EventDispatcher()
    let onPageChange = EventDispatcher()
    let onBookLoaded = EventDispatcher()
    let onMiddleTouch = EventDispatcher()
    let onError = EventDispatcher()
    
    public var navigator: PDFNavigatorViewController?
    
    public var pendingProps: PDFProps = .init()
    public var props: FinalizedPDFProps?
    
    private var isInitialized = false
    
    // Tasks for cleanup
    private var loadPublicationTask: Task<Void, Never>?
    private var positionsTask: Task<Void, Never>?
    private var navigationTasks: [Task<Void, Never>] = []
    
    // Background handling
    private var backgroundObserver: NSObjectProtocol?
    private var foregroundObserver: NSObjectProtocol?
    private var isInBackground = false
    
    public required init(appContext: AppContext? = nil) {
        super.init(appContext: appContext)
        setupBackgroundObservers()
    }
    
    deinit {
        print("PDFView: deinit called - cleaning up resources")
        cancelAllTasks()
        removeBackgroundObservers()
        cleanupNavigator()
    }
    
    private func setupBackgroundObservers() {
        backgroundObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppDidEnterBackground()
        }
        
        foregroundObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.handleAppWillEnterForeground()
        }
    }
    
    private func removeBackgroundObservers() {
        if let observer = backgroundObserver {
            NotificationCenter.default.removeObserver(observer)
            backgroundObserver = nil
        }
        if let observer = foregroundObserver {
            NotificationCenter.default.removeObserver(observer)
            foregroundObserver = nil
        }
    }
    
    private func handleAppDidEnterBackground() {
        print("PDFView: App entered background - suspending operations")
        isInBackground = true
        positionsTask?.cancel()
    }
    
    private func handleAppWillEnterForeground() {
        print("PDFView: App entering foreground - resuming operations")
        isInBackground = false
    }
    
    private func cancelAllTasks() {
        print("PDFView: Cancelling all tasks")
        loadPublicationTask?.cancel()
        positionsTask?.cancel()
        navigationTasks.forEach { $0.cancel() }
        navigationTasks.removeAll()
    }
    
    private func cleanupNavigator() {
        print("PDFView: Cleaning up navigator")
        navigator?.view.removeFromSuperview()
        navigator = nil
    }
    
    public func finalizeProps() {
        let oldProps = props
        
        // Don't proceed if we don't have required props
        guard let bookId = pendingProps.bookId,
              let url = pendingProps.url
        else {
            return
        }
        
        props = FinalizedPDFProps(
            bookId: bookId,
            locator: pendingProps.locator ?? pendingProps.initialLocator ?? oldProps?.locator,
            url: url,
            background: pendingProps.background ?? oldProps?.background ?? Color(hex: "#000000")!,
            pageSpacing: pendingProps.pageSpacing ?? oldProps?.pageSpacing ?? 0.0,
            scrollAxis: pendingProps.scrollAxis ?? oldProps?.scrollAxis ?? .vertical,
            scroll: pendingProps.scroll ?? oldProps?.scroll ?? true,
            readingProgression: pendingProps.readingProgression ?? oldProps?.readingProgression ?? .ltr,
            spread: pendingProps.spread ?? oldProps?.spread ?? .auto
        )
        
        
        // If this is a new book or first initialization, load the publication
        if props!.bookId != oldProps?.bookId || props!.url != oldProps?.url || !isInitialized {
            loadPublicationTask?.cancel()
            loadPublicationTask = Task { [weak self] in
                await self?.loadPublication()
            }
            return
        }
        
        // Update navigator if locator changed
        if props!.locator != oldProps?.locator, let locator = props!.locator {
            go(locator: locator)
        }
        
        if isInitialized && preferencesChanged(oldProps: oldProps) {
            updatePreferences()
        }
    }
    
    private func preferencesChanged(oldProps: FinalizedPDFProps?) -> Bool {
        guard let oldProps = oldProps, let props = props else { return false }
        return props.background != oldProps.background ||
               props.pageSpacing != oldProps.pageSpacing ||
               props.scrollAxis != oldProps.scrollAxis ||
               props.scroll != oldProps.scroll ||
               props.readingProgression != oldProps.readingProgression ||
               props.spread != oldProps.spread
    }
    
    private func loadPublication() async {
        guard let props = props else { return }
        
        do {
            if let url = URL(string: props.url) {
                var publicationUrl = url
                
                if url.scheme == "http" || url.scheme == "https" {
                    publicationUrl = try await downloadPDF(from: url)
                }
                
                let publication = try await BookService.instance.openPublication(for: props.bookId, at: publicationUrl)
                
                try Task.checkCancellation()
                
                await MainActor.run { [weak self] in
                    self?.initializeNavigator(with: publication)
                }
            }
        } catch {
            if error is CancellationError {
                print("Publication load cancelled")
                return
            }
            
            print("Error loading publication: \(error)")
            await MainActor.run { [weak self] in
                self?.onError([
                    "errorDescription": error.localizedDescription,
                    "failureReason": "Failed to load PDF",
                    "recoverySuggestion": "Check the URL and try again",
                ])
            }
        }
    }
    
    private func downloadPDF(from url: URL) async throws -> URL {
        let (data, _) = try await URLSession.shared.data(from: url)
        
        let tempDirectory = FileManager.default.temporaryDirectory
        let pdfFile = tempDirectory.appendingPathComponent(UUID().uuidString + ".pdf")
        
        try data.write(to: pdfFile)
        return pdfFile
    }
    
    public func initializeNavigator(with publication: Publication) {
        guard let props = props else { return }
        
        do {
            let navigator = try PDFNavigatorViewController(
                publication: publication,
                initialLocation: props.locator,
                config: .init(
                    preferences: PDFPreferences(
                        backgroundColor: props.background,
                        pageSpacing: props.pageSpacing,
                        readingProgression: props.readingProgression,
                        scroll: props.scroll,
                        scrollAxis: props.scrollAxis,
                        spread: props.spread,
                        visibleScrollbar: false
                    )
                ),
                httpServer: GCDHTTPServer(
                    assetRetriever: AssetRetriever(httpClient: DefaultHTTPClient())
                )
            )
            
            navigator.delegate = self
            addSubview(navigator.view)
            self.navigator = navigator
            isInitialized = true
            
            positionsTask?.cancel()
            positionsTask = Task { [weak self] in
                guard let self = self else { return }
                
                let positionsResult = await publication.positions()
                let totalPages = (try? positionsResult.get().count) ?? 0
                
                try? Task.checkCancellation()
                
                await MainActor.run { [weak self] in
                    self?.onBookLoaded([
                        "success": true,
                        "bookMetadata": [
                            "title": publication.metadata.title ?? "",
                            "author": publication.metadata.authors.map { $0.name }.joined(separator: ", "),
                            "publisher": publication.metadata.publishers.map { $0.name }.joined(separator: ", "),
                            "identifier": publication.metadata.identifier ?? "",
                            "language": publication.metadata.languages.first ?? "en",
                            "totalPages": totalPages,
                        ],
                    ])
                }
            }
            
            emitCurrentLocator()
            
        } catch {
            print("Failed to create PDF Navigator instance: \(error)")
            onError([
                "errorDescription": error.localizedDescription,
                "failureReason": "Failed to create PDF navigator",
                "recoverySuggestion": "Try reloading the PDF",
            ])
        }
    }
    
    public func destroyNavigator() {
        print("PDFView: destroyNavigator called")
        
        cancelAllTasks()
        
        navigator?.view.removeFromSuperview()
        navigator = nil
        isInitialized = false
        
        if let bookId = props?.bookId {
            BookService.instance.closePublication(for: bookId)
        }
    }
    
    // TODO: Determine if I even need a locator for PDFs + Readium, just using page numbers
    // would be much less hassle
    func emitCurrentLocator() {
        guard let navigator = navigator,
              let currentLocator = navigator.currentLocation
        else {
            return
        }
        
        onLocatorChange(makeJSON([
            "href": currentLocator.href.string,
            "title": encodeIfNotNil(currentLocator.title),
            "locations": encodeIfNotEmpty(currentLocator.locations.json),
            "text": encodeIfNotEmpty(currentLocator.text.json),
            "type": encodeIfNotEmpty(currentLocator.mediaType.string),
        ]))
        
        if let pageNumber = currentLocator.locations.position {
            onPageChange([
                "currentPage": pageNumber,
            ])
        }
    }
    
    func go(locator: Locator) {
        let task = Task { [weak self] in
            guard let self = self else { return }
            _ = await self.navigator?.go(to: locator, options: NavigatorGoOptions(animated: true))
        }
        navigationTasks.append(task)
        navigationTasks.removeAll { $0.isCancelled }
    }
    
    func goToLocation(locator: Locator) {
        go(locator: locator)
    }
    
    func goToPage(page: Int) {
        guard let navigator = navigator else { return }
        
        let task = Task { [weak self] in
            guard let self = self else { return }
            
            let positionsResult = await navigator.publication.positions()
            guard let positions = try? positionsResult.get() else { return }
            
            // Ensure page is within bounds
            guard page > 0 && page <= positions.count else {
                print("PDFView: Invalid page number \(page)")
                return
            }
            
            let locator = positions[page - 1]
            _ = await navigator.go(to: locator, options: NavigatorGoOptions(animated: true))
        }
        navigationTasks.append(task)
        navigationTasks.removeAll { $0.isCancelled }
    }
    
    func goForward() {
        let task = Task { [weak self] in
            guard let self = self else { return }
            _ = await self.navigator?.goForward(options: NavigatorGoOptions(animated: true))
        }
        navigationTasks.append(task)
        navigationTasks.removeAll { $0.isCancelled }
    }
    
    func goBackward() {
        let task = Task { [weak self] in
            guard let self = self else { return }
            _ = await self.navigator?.goBackward(options: NavigatorGoOptions(animated: true))
        }
        navigationTasks.append(task)
        navigationTasks.removeAll { $0.isCancelled }
    }
    
    func updatePreferences() {
        guard let props = props else { return }
        
        let preferences = PDFPreferences(
            backgroundColor: props.background,
            pageSpacing: props.pageSpacing,
            readingProgression: props.readingProgression,
            scroll: props.scroll,
            scrollAxis: props.scrollAxis,
            spread: props.spread
        )
        
        navigator?.submitPreferences(preferences)
    }
    
    override public func layoutSubviews() {
        super.layoutSubviews()
        guard let navigatorView = navigator?.view else {
            return
        }
        navigatorView.frame = bounds
    }
}

extension PDFView: PDFNavigatorDelegate {
    public func navigator(_: Navigator, locationDidChange _: Locator) {
        emitCurrentLocator()
    }
    
    public func navigator(_: Navigator, presentError error: NavigatorError) {
        onError([
            "errorDescription": error.localizedDescription,
            "failureReason": "Navigation failed",
            "recoverySuggestion": "Try again",
        ])
    }
    
    public func navigator(_ navigator: VisualNavigator, didTapAt point: CGPoint) {
        let navigator = navigator as! PDFNavigatorViewController
        
        if point.x < bounds.maxX * 0.2 {
            let task = Task { [weak self] in
                guard self != nil else { return }
                _ = await navigator.goBackward(options: NavigatorGoOptions(animated: true))
            }
            navigationTasks.append(task)
            navigationTasks.removeAll { $0.isCancelled }
            return
        }
        if point.x > bounds.maxX * 0.8 {
            let task = Task { [weak self] in
                guard self != nil else { return }
                _ = await navigator.goForward(options: NavigatorGoOptions(animated: true))
            }
            navigationTasks.append(task)
            navigationTasks.removeAll { $0.isCancelled }
            return
        }
        
        onMiddleTouch()
    }
}

File: apps/expo/modules/readium/android/src/main/java/expo/modules/readium/PDFView.kt
@file:OptIn(ExperimentalReadiumApi::class)

package expo.modules.readium

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.PointF
import android.util.Log
import android.view.View
import android.widget.FrameLayout
import android.widget.TextView
import androidx.annotation.ColorInt
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.commitNow
import androidx.lifecycle.lifecycleScope
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import kotlinx.coroutines.*
import org.readium.r2.navigator.DecorableNavigator
import org.readium.r2.navigator.input.InputListener
import org.readium.r2.navigator.input.TapEvent
import org.readium.r2.navigator.pdf.PdfNavigatorFragment
import org.readium.r2.shared.ExperimentalReadiumApi
import org.readium.r2.shared.InternalReadiumApi
import org.readium.r2.shared.extensions.toMap
import org.readium.r2.shared.publication.Publication
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.publication.ReadingProgression
import org.readium.r2.shared.publication.services.positions
import java.net.URL

data class PDFProps(
    var bookId: String? = null,
    var locator: Locator? = null,
    var initialLocator: Locator? = null,
    var url: String? = null,
    @ColorInt var background: Int? = null,
    var pageSpacing: Double? = null,
    var scrollAxis: String? = null,
    var scroll: Boolean? = null,
    var readingProgression: ReadingProgression? = null,
    var spread: org.readium.r2.navigator.preferences.Spread? = null
)

data class FinalizedPDFProps(
    val bookId: String,
    val locator: Locator?,
    val url: String,
    @ColorInt val background: Int,
    val pageSpacing: Double,
    val scrollAxis: String,
    val scroll: Boolean,
    val readingProgression: ReadingProgression,
    val spread: org.readium.r2.navigator.preferences.Spread
)

@SuppressLint("ViewConstructor", "ResourceType")
class PDFView(context: Context, appContext: AppContext) : ExpoView(context, appContext),
    DecorableNavigator.Listener, PdfNavigatorFragment.Listener {

    // Required for proper layout! Forces Expo to
    // use the Android layout system for this view,
    // rather than React Native's. Without this,
    // the ViewPager and WebViews will be laid out
    // incorrectly
    override val shouldUseAndroidLayout = true

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        Log.d("PDFView", "onLayout: changed=$changed, dimensions=${right-left}x${bottom-top}")
    }

    var bookService: BookService? = null

    val onLocatorChange by EventDispatcher()
    val onPageChange by EventDispatcher()
    val onBookLoaded by EventDispatcher()
    val onMiddleTouch by EventDispatcher()
    val onError by EventDispatcher()

    var navigator: PdfNavigatorFragment<*, *>? = null
    private var publication: Publication? = null
    private var pdfFragment: PDFFragment? = null

    val pendingProps = PDFProps()
    var props: FinalizedPDFProps? = null

    private val placeholderView = TextView(context).apply {
        layoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        ).apply {
            gravity = android.view.Gravity.CENTER
        }
//        TODO: Spinner
        text = "Loading PDF..."
        textAlignment = View.TEXT_ALIGNMENT_CENTER
        setTextColor(Color.BLACK)
    }

    private var coroutineScope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var locatorCollectionJob: Job? = null

    fun finalizeProps() {
        val oldProps = props

        val bookId = pendingProps.bookId ?: return
        val url = pendingProps.url ?: return

        props = FinalizedPDFProps(
            bookId = bookId,
            locator = pendingProps.locator ?: pendingProps.initialLocator ?: oldProps?.locator,
            url = url,
            background = pendingProps.background
                ?: oldProps?.background ?: Color.parseColor("#000000"),
            pageSpacing = pendingProps.pageSpacing ?: oldProps?.pageSpacing ?: 0.0,
            scrollAxis = pendingProps.scrollAxis ?: oldProps?.scrollAxis ?: "vertical",
            scroll = pendingProps.scroll ?: oldProps?.scroll ?: true,
            readingProgression = pendingProps.readingProgression 
                ?: oldProps?.readingProgression ?: ReadingProgression.LTR,
            spread = pendingProps.spread 
                ?: oldProps?.spread ?: org.readium.r2.navigator.preferences.Spread.AUTO
        )

        // If book ID or URL changed, reload the publication
        if (props!!.bookId != oldProps?.bookId || props!!.url != oldProps?.url) {
            Log.d("PDFView", "Book ID or URL changed, reloading publication")
            destroyNavigator()
            coroutineScope.launch {
                loadPublication()
            }
            return
        }

        // Check if any preferences changed
        val prefsChanged = oldProps != null && (
            props!!.pageSpacing != oldProps.pageSpacing ||
            props!!.scrollAxis != oldProps.scrollAxis ||
            props!!.scroll != oldProps.scroll ||
            props!!.readingProgression != oldProps.readingProgression ||
            props!!.spread != oldProps.spread
        )

        if (prefsChanged) {
            destroyNavigator()
            coroutineScope.launch {
                loadPublication()
            }
            return
        }

        // If locator changed, navigate to it
        if (props!!.locator != oldProps?.locator && props!!.locator != null) {
            go(props!!.locator!!)
        }
    }

    private suspend fun loadPublication() {
        val props = props ?: return
        
        Log.d("PDFView", "Loading publication from: ${props.url}")
        
        withContext(Dispatchers.Main) {
            addView(placeholderView)
        }

        try {
            val url = URL(props.url)
            bookService?.openPublication(props.bookId, url)
            
            withContext(Dispatchers.Main) {
                initializeNavigator()
                removeView(placeholderView)
            }
        } catch (e: Exception) {
            Log.e("PDFView", "Failed to load publication", e)
            withContext(Dispatchers.Main) {
                removeView(placeholderView)
                onError(
                    mapOf(
                        "errorDescription" to (e.message ?: "Unknown error"),
                        "failureReason" to "Failed to load PDF",
                        "recoverySuggestion" to "Check the URL and try again"
                    )
                )
            }
        }
    }

    fun initializeNavigator() {
        val publication = bookService?.getPublication(props!!.bookId) ?: run {
            Log.e("PDFView", "Publication not found for bookId: ${props!!.bookId}")
            return
        }

        Log.d("PDFView", "Publication loaded successfully: ${publication.metadata.title}")

        val fragmentTag = resources.getString(R.string.pdf_fragment_tag)
        val activity: FragmentActivity? = appContext.currentActivity as FragmentActivity?

        if (activity == null) {
            Log.e("PDFView", "Cannot initialize navigator: activity is null")
            return
        }

        val listener = this
        pdfFragment = PDFFragment(
            publication,
            props!!,
            listener
        )

        Log.d("PDFView", "Adding PDFFragment to activity FragmentManager")
        activity.supportFragmentManager.commitNow {
            setReorderingAllowed(true)
            add(pdfFragment!!, fragmentTag)
        }
        
        addView(pdfFragment!!.view, FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.MATCH_PARENT
        ))
        
        post {
            val fragment = pdfFragment ?: run {
                Log.e("PDFView", "[POST] pdfFragment is null!?")
                return@post
            }
            
            navigator = fragment.navigator
            
            if (navigator == null) {
                Log.e("PDFView", "[POST] Navigator is null!?")
                return@post
            }
            
            setupNavigatorListeners(publication)
            
            // Navigate to initial locator if provided
            props?.locator?.let { locator ->
                go(locator)
            }
        }
    }

    private fun setupNavigatorListeners(publication: Publication) {
        navigator?.addInputListener(TapInputListener())
        
        locatorCollectionJob = coroutineScope.launch {
            try {
                navigator?.currentLocator?.collect { locator ->
                    if (isAttachedToWindow) {
                        emitCurrentLocator(locator)
                    }
                }
            } catch (e: Exception) {
                Log.e("PDFView", "Error collecting locator", e)
            }
        }

        coroutineScope.launch {
            try {
                if (isAttachedToWindow) {
                    emitBookLoaded(publication)
                }
            } catch (e: Exception) {
                Log.e("PDFView", "Error emitting book loaded", e)
            }
        }
    }

    fun destroyNavigator() {
        Log.d("PDFView", "destroyNavigator called")
        
        val fragment = this.pdfFragment ?: run {
            Log.d("PDFView", "Navigator already destroyed")
            return
        }
        
        this.navigator = null
        this.pdfFragment = null
        
        locatorCollectionJob?.cancel()
        locatorCollectionJob = null
        
        // Remove the fragment's view from PDFView
        try {
            removeView(fragment.view)
        } catch (e: Exception) {
            Log.e("PDFView", "Error removing fragment view", e)
        }
        
        val activity: FragmentActivity? = appContext.currentActivity as? FragmentActivity
        if (activity == null || activity.isDestroyed || activity.isFinishing) {
            Log.w("PDFView", "Cannot remove fragment: activity is gone")
        } else {
            try {
                activity.supportFragmentManager.commitNow {
                    setReorderingAllowed(true)
                    remove(fragment)
                }
            } catch (e: IllegalStateException) {
                Log.e("PDFView", "Failed to remove fragment: ${e.message}")
            }
        }
        
        val bookId = props?.bookId
        if (bookId != null) {
            bookService?.closePublication(bookId)
        }
    }

    @OptIn(InternalReadiumApi::class)
    private fun emitCurrentLocator(locator: Locator) {
        onLocatorChange(locator.toJSON().toMap())
        
        val pageIndex = locator.locations.position ?: 1
        onPageChange(
            mapOf(
                "page" to pageIndex,
                "progression" to (locator.locations.progression ?: 0.0)
            )
        )
    }

    private suspend fun emitBookLoaded(publication: Publication) {
        val positions = publication.positions()
        val totalPages = positions.size
        
        onBookLoaded(
            mapOf(
                "success" to true,
                "bookMetadata" to mapOf(
                    "title" to (publication.metadata.title ?: ""),
                    "author" to publication.metadata.authors.joinToString(", ") { it.name.toString() },
                    "publisher" to publication.metadata.publishers.joinToString(", ") { it.name.toString() },
                    "identifier" to (publication.metadata.identifier ?: ""),
                    "language" to (publication.metadata.languages.firstOrNull() ?: "en"),
                    "totalPages" to totalPages
                )
            )
        )
    }

    // Navigation methods
    fun go(locator: Locator) {
        navigator?.go(locator, animated = true)
    }

    fun goToLocation(locator: Locator) {
        go(locator)
    }

    fun goToPage(page: Int) {
        coroutineScope.launch {
            try {
                val publication = bookService?.getPublication(props!!.bookId) ?: return@launch
                val positions = publication.positions()
                
                // Ensure page is within bounds
                if (page <= 0 || page > positions.size) {
                    Log.e("PDFView", "Invalid page number: $page")
                    return@launch
                }
                
                val locator = positions[page - 1]
                navigator?.go(locator, animated = true)
            } catch (e: Exception) {
                Log.e("PDFView", "Error navigating to page", e)
            }
        }
    }

    fun goForward() {
        navigator?.goForward(animated = true)
    }

    fun goBackward() {
        navigator?.goBackward(animated = true)
    }

    // Handle tap events for navigation and middle touch
    fun handleTap(point: PointF): Boolean {
        if (point.x < width * 0.2) {
            navigator?.goBackward(animated = true)
            return true
        }
        if (point.x > width * 0.8) {
            navigator?.goForward(animated = true)
            return true
        }
        onMiddleTouch(mapOf())
        return false
    }

    // TODO: Support decoration events
    override fun onDecorationActivated(event: DecorableNavigator.OnActivatedEvent): Boolean {
        return false
    }

    /**
     * Input listener to handle tap events for navigation
     */
    private inner class TapInputListener : InputListener {
        override fun onTap(event: TapEvent): Boolean {
            return handleTap(event.point)
        }
    }
}
